grr/server/grr_response_server/flows/general/osquery_test.py

```diff
@@ -148,47 +153,101 @@ def testFailure(self):
+      OrderedDict([ ("foo", "blargh"), ("bar", "plugh"), ("baz", "ztesch") ]),
     ]
-    """
+    stdout = json.Dump(test_table)
+  
     with osquery_test_lib.FakeOsqueryiOutput(stdout=stdout, stderr=""):
       results = self._RunQuery("SELECT foo, bar, baz FROM foobarbaz;")
 
     self.assertLen(results, 1)
 
     table = results[0].table
-    self.assertLen(table.header.columns, 3)
-    self.assertEqual(table.header.columns[0].name, "foo")
-    self.assertEqual(table.header.columns[1].name, "bar")
-    self.assertEqual(table.header.columns[2].name, "baz")
-    self.assertEqual(list(table.Column("foo")), ["quux", "blargh"])
-    self.assertEqual(list(table.Column("bar")), ["norf", "plugh"])
-    self.assertEqual(list(table.Column("baz")), ["thud", "ztesch"])
+    self._AssertTablesMatch(table, test_table)
 
   def testFailure(self):
     stderr = "Error: near '*': syntax error"
     with osquery_test_lib.FakeOsqueryiOutput(stdout="", stderr=stderr):
       with self.assertRaises(RuntimeError):
         self._RunQuery("SELECT * FROM *;")
 
+  def testProgressDoesntTruncate10Rows(self):
+    query = "doesn't matter"
+    number_of_columns = 5
+    number_of_rows = 10
+
+    test_table = _CreateDictTable(number_of_columns, number_of_rows)
+    test_table_json = json.Dump(test_table)
+
+    with osquery_test_lib.FakeOsqueryiOutput(stdout=test_table_json, stderr=""):
+      flow_id = self._NewTestFlowStartAndRun(query)
+      progress: rdf_osquery.OsqueryProgress = flow_test_lib.GetFlowProgress(self.client_id, flow_id)
+
+    self.assertEqual(progress.total_rows_count, number_of_rows)
+    self.assertEqual(progress.partial_table.query, query)
+    self._AssertTablesMatch(progress.partial_table, test_table)
+
+  def testProgressTruncates20RowsTo10(self):
+    query = "doesn't matter"
+    number_of_columns = 5
+    number_of_rows = 20
+
+    test_table = _CreateDictTable(number_of_columns, number_of_rows)
+    test_table_json = json.Dump(test_table)
+
+    with osquery_test_lib.FakeOsqueryiOutput(stdout=test_table_json, stderr=""):
+      flow_id = self._NewTestFlowStartAndRun(query)
+      progress: rdf_osquery.OsqueryProgress = flow_test_lib.GetFlowProgress(self.client_id, flow_id)
+
+    self.assertEqual(progress.total_rows_count, number_of_rows)
+    self.assertEqual(progress.partial_table.query, query)
+    self._AssertTablesMatch(progress.partial_table, test_table[:10])
+
+  def _NewTestFlowStartAndRun(self, query):
+    session_id = flow_test_lib.TestFlowHelper(
+        osquery_flow.OsqueryFlow.__name__,
+        action_mocks.ActionMock(osquery_action.Osquery),
+        client_id=self.client_id,
+        token=self.token,
+        query=query)
+    return session_id
+
+  def _RunQuery(self, query):
+    session_id = self._NewTestFlowStartAndRun(query)
+    return flow_test_lib.GetFlowResults(self.client_id, session_id)
+
+  def _AssertTablesMatch(self, table: rdf_osquery.OsqueryTable, expected: List[OrderedDict]) -> None:
```

Comment:
I agree in this particular case - my code should be consistent with the rest of the code base.

In general, though, what are the benefits of that? I guess this might be a left-over from languages like C in which you need to declare things before using them.

If instead protected methods are put at the end, they will make way for the public methods to be first. The public methods are the interface exposed, so if anyone needs to look at the class they will be presented with the core high-level functionality first. This in my opinion is better than first exposing low-level helper functions?

Reply:
> If instead protected methods are put at the end, they will make way for the public methods to be first. The public methods are the interface exposed, so if anyone needs to look at the class they will be presented with the core high-level functionality first. This in my opinion is better than first exposing low-level helper functions?

This is a good question. I don't have a strong opinion on this as I think that practical differences between different ways of grouping methods together are negligible. It's normally useful to have them grouped (even though sometimes it actually makes sense to put a protected method right next to the public method using it), but whether it's better to have public first, protected second or protected first, public second - feels like a subjective preference. Developers have different backgrounds, use different IDEs, etc.

I personally find it useful to have the implementation details grouped together, and since often private object properties are defined at the top of the class, I see the point of grouping them with private/protected methods. Then, when I read the class definition I skim through all the private/protected stuff, getting an idea of what is there, and then I can read through public implementation with that additional context in mind. Again, this is very subjective, so the only real guideline in this question is code consistency - whichever approach we choose, makes sense to stick to it.

Reply:
> Then, when I read the class definition I skim through all the private/protected stuff, getting an idea of what is there, and then I can read through public implementation with that additional context in mind.

Here, I agree with @simstoykov. Why should you need to read the private and protected parts of the class? Public interface is all you in general should care about. Sometimes it is not (if the code is poorly documented), but we should write code with assumption that we write good code, not a bad one.

Reply:
IMHO, both approaches make sense, both "public members first" and "c-style bottom-up definition". Since we do not have a guideline or style guide regarding this, I would favour accepting any sensible grouping of class members in this PR.